Add a parameter object on the system, use it in derived classes

GiudGiud opened the pull request:
refs #4006

If this is too much, I could scale this down to just the nonlinear_implicit_system. That's all I need. 
We could consider a getParam routine that does this 3-liner operation instead.

roystgnr commented on src/systems/frequency_system.C:
```diff
@@ -112,14 +112,18 @@ void FrequencySystem::init_data ()
     {
+      // not supported for now
+      if (parameters.have_parameter<unsigned int> ("n_frequencies"))
+        libmesh_error_msg("ERROR: Not supported");
```
Why not?  Seems like a straightforward behavior even if we lack the proper test coverage for it.

GiudGiud replied:
there's quite a few places to make changes to make it work

GiudGiud replied:
like in `set_frequencies_by_range` do we set the es.parameters now, or switch to the local system, or both?

roystgnr replied:
Okay, that's a good reason.  Setting just local system parameters breaks backwards compatibility, setting just es means we might not actually change the parameters used, and setting both means we lose the new flexibility.  I'd say change this to `libmesh_not_implemented_msg()` but there's no need to actually change the behavior now.

roystgnr commented on src/systems/frequency_system.C:
```diff
@@ -112,14 +112,18 @@ void FrequencySystem::init_data ()
 #ifndef NDEBUG
           const unsigned int n_freq =
-            es.parameters.get<unsigned int>("n_frequencies");
+            es.para) :<unsigned int>("n_frequencies");
```
Looks like a search-and-replace went badly awry somewhere.

GiudGiud replied:
either that or switching tabs then starting typing in the console but it's still typing on the text editor

GiudGiud replied:
error was in two places so def search and replace

roystgnr commented:
I like this, and I wouldn't scale it down, though I'm crossing my fingers hoping we aren't breaking any user subclasses that also thought `parameters` was a good name for a local variable.

GiudGiud commented:
@roystgnr do you have thoughts on testing this? Do you want to rely on our use in MOOSE for testing? This will be used for new Convergence objects for each nonlinear system in a multi-system context.

Or should there be an example? Or some other form of testing?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2550046/), step Generate coverage on 4d281ca wanted to post the following:


## Coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>029330</th>
      <th colspan=4>#4007 4d281c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>62.35%</td>
      <td>62.34%</td>
      <td>-0.01%</td>
      <td>37.88%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72649</td>
      <td>72671</td>
      <td>+22</td>
      <td>50</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>43865</td>
      <td>43892</td>
      <td>+27</td>
      <td>82</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/4007/coverage_diff)

[Full coverage report](https://mooseframework.inl.gov/libmesh/docs/PRs/4007/coverage)

## Warnings

- New new line coverage rate 37.88% is less than the suggested 90.0%



This comment will be updated on new commits.

roystgnr commented:
I hate relying on MOOSE tests for coverage, and IIRC we deliberately exclude MOOSE tests from our coverage stats for that reason.  It's hard to come up with a good unit test here, but ... maybe set up a problem that should require a lot of linear iterations, but then set a system parameter that holds us to a tiny max number of iterations, and check the number of iterations reported after a `solve()`?  We can't assert that it's *equal* to our max iterations (unless we set max=1? that wouldn't be crazy) lest we break with `-pc_type lu`, but we can at least assert that it doesn't exceed our max iterations.

GiudGiud commented:
I added a test that should show that parameters are read from the system rather than the equation system when both are present. It's not the best either because if we changed the name of the parameter of the system, the test would pass even through the change does nothing. 
If we want 100% coverage we'll need a test for every system and every parameter?

GiudGiud commented:
ok made a little mistake. going to push it again soon

roystgnr commented on src/systems/frequency_system.C:
```diff
@@ -112,14 +112,18 @@ void FrequencySystem::init_data ()
 #ifndef NDEBUG
           const unsigned int n_freq =
-            es.parameters.get<unsigned int>("n_frequencies");
+            es.parameters:<unsigned int>("n_frequencies");
```
Another search-and-replace bug?

roystgnr commented on src/systems/frequency_system.C:
```diff
@@ -304,7 +308,7 @@ void FrequencySystem::set_frequencies (const std::vector<Number> & frequencies,
 {
   libmesh_assert(_finished_set_frequencies);
-  return this->get_equation_systems().parameters.get<unsigned int>("n_frequencies");
+  return this->get_equation_systems().parameters:<unsigned int>("n_frequencies");
```
And here

roystgnr commented on src/systems/implicit_system.C:
```diff
@@ -1222,8 +1222,16 @@ LinearSolver<Number> * ImplicitSystem::get_linear_solver() const
+    return std::make_pair(this->get_equation_systems().parameters.get<unsigned int>("linear solver maximum iterations"),
+                          this->get_equation_systems().parameters.get<Real>("linear solver tolerance"));
+  else
+    libmesh_error_msg("ERROR: Insufficient linear solver parameters");
```
This isn't what we want to do, is it?  If we override one parameter for a particular system, we should still just default to the EquationSystems parameter value for the other.

GiudGiud replied:
ah that works.
I was thinking of forcing the user to be consistent between where each parameter comes from

it s probably more flexible this other way

roystgnr commented on src/systems/frequency_system.C:
```diff
@@ -339,9 +343,11 @@ void FrequencySystem::solve (const unsigned int n_start,
   //     the user-specified wave speed
-  const Real tol            =
+  const Real tol            = parameters.have_parameter<Real>("linear solver tolerance") ?
+    double(parameters.get<Real>("linear solver tolerance")) :
     es.parameters.get<Real>("linear solver tolerance");
-  const unsigned int maxits =
+  const unsigned int maxits = parameters.have_parameter<unsigned int>("linear solver maximum iterations") ?
+    parameters.get<unsigned int>("linear solver maximum iterations") :
     es.parameters.get<unsigned int>("linear solver maximum iterations");
```
As long as we have to change all these anyway, wouldn't `auto [maxits, tol] = this->get_linear_solve_parameters();` work?  We can't do that in `EigenSystem` because it doesn't inherit from `ImplicitSystem`, but it seems like it ought to work here, and in `LinearImplicitSystem`, and (albeit for just these two parameters) in `NonlinearImplicitSystem`.

roystgnr commented on tests/systems/systems_test.C:
```diff
@@ -1256,6 +1257,78 @@ public:
+
+    // We set the solution to be 1 everywhere, so the final l1 norm of the
+    // solution is the product of the number of variables and number of nodes.
+    Real ref_l1_norm = static_cast<Real>(mesh.n_nodes() * li_system.n_vars());
```
Instead of using enough iterations to converge and then testing convergence, we should be using too few iterations to converge (so we stop at the max iteration count) and testing to make sure we're at the specified max iteration count.

GiudGiud replied:
will do

roystgnr commented on tests/systems/systems_test.C:
```diff
@@ -1256,6 +1257,78 @@ public:
   }
 
+
+  void testSetSystemParameterOverEquationSystem()
```
This test isn't currently being run.  You need to add a `CPPUNIT_TEST( testSystemParameterOverEquationSystems );` macro invocation with the others at the start of the class.

roystgnr commented:
We might be out of luck for the nonlinear test; that has no way to tell you the final step's linear iteration count, only the nonlinear iteration count.  The linear test ought to be sufficient, except that you'll have to use `5u` instead of `5` there, because cppunit isn't clever enough to cope with tests that mix different types, even as obviously-mixable types as `int` and `unsigned int`.

GiudGiud commented:
for the nonlienar system I kept my old test essentially, checking another parameter rather than num linear iterations

GiudGiud commented:
@roystgnr when you have a minute.
The plan after this is to have a NonlinearResidual convergence object in moose on a per system basis

roystgnr commented:
Thanks for the extra ping; not sure how I missed the review request last week, but I'm glad I got this in time to squeeze it into the submodule update PR tonight.
